log/log.go

```diff
@@ -0,0 +1,180 @@
+// Copyright © 2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strings"
+)
+
+// Specifies the log levels
+const (
+	LevelError = iota + 1
+	LevelWarning
+	LevelInfo
+	LevelDebug
+	LevelVerbose // This is the same as Debug, but without the repeat filtering
+)
+
+// Specifies the command tags
+const (
+	Mixer       = "MIXER"
+	Dnf         = "DNF"
+	Rpm         = "RPM"
+	Rpm2Archive = "RPM2ARCHIVE"
+	Tar         = "TAR"
+	BsDiff      = "BSDIFF"
+)
+
+var (
+	level      = LevelInfo
+	levelMap   = map[int]string{}
+	fileHandle *os.File
+	logging    = false
+	lineLast   string
+	lineCount  int
+	cmdMap     = map[string]bool{}
+)
+
+func init() {
+	levelMap[LevelError] = "ERROR"
+	levelMap[LevelWarning] = "WARNING"
+	levelMap[LevelInfo] = "INFO"
+	levelMap[LevelDebug] = "DEBUG"
+	levelMap[LevelVerbose] = "VERBOSE"
+	cmdMap[Mixer] = true
+	cmdMap[Dnf] = true
+	cmdMap[Rpm] = true
+	cmdMap[Rpm2Archive] = true
+	cmdMap[Tar] = true
+	cmdMap[BsDiff] = true
+}
+
+// SetLogLevel sets the default log level to l
+func SetLogLevel(l int) {
+	if l < LevelError {
+		level = LevelError
+		logTag("WRN", Mixer, "Log Level '%d' too low, forcing to %s (%d)", l, levelMap[level], level)
+	} else if l > LevelVerbose {
+		level = LevelVerbose
+		logTag("WRN", Mixer, "Log Level '%d' too high, forcing to %s (%d)", l, levelMap[level], level)
+	} else {
+		level = l
+		Debug(Mixer, "Log Level set to %s (%d)", levelMap[level], l)
+	}
+}
+
+// SetOutputFilename ... sets the default log output to filename instead of stdout/stderr
+func SetOutputFilename(logFile string) (*os.File, error) {
+	var err error
+	fileHandle, err = os.OpenFile(logFile, os.O_RDWR|os.O_APPEND|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, err
+	}
+
+	log.SetOutput(fileHandle)
+	logging = true
+	return fileHandle, nil
+}
+
+// GetLogHandler ... returns the file handler of logging file
+func GetLogHandler() *os.File {
+	return fileHandle
+}
+
+func logTag(tag string, cmdTag, format string, a ...interface{}) {
+	// If there are no variable to pass to the format,
+	// then we can escape any % signs.
+	if len(a) < 1 {
+		format = strings.ReplaceAll(format, "%", "%%")
+	}
+
+	f := "[" + tag + "] " + "[" + cmdTag + "] " + format + "\n"
```

Comment:
Lets remove the space between annotations. E.g.
`"[" + tag + "]" + "[" + cmdTag + "]"`